rp2040-hal/src/pll.rs

```diff
@@ -0,0 +1,254 @@
+//! Phase-Locked Loops (PLL)
+// See [Chapter 2 Section 18](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
+
+use core::{
+    convert::{
+        Infallible,
+        TryFrom,
+        TryInto
+    },
+    marker::PhantomData,
+    ops::{
+        RangeInclusive,
+        Range,
+        Deref
+    }
+};
+
+use embedded_time::{
+    fixed_point::FixedPoint,
+    rate::{
+        Hertz,
+        Generic,
+        Rate
+    }
+};
+
+use nb::Error::WouldBlock;
+
+/// State of the PLL
+pub trait State {}
+
+
+/// PLL is disabled.
+pub struct Disabled;
+
+/// PLL is configured, started and locking into its designated frequency.
+pub struct Locking {
+    post_div1: u8,
+    post_div2: u8
+}
+
+/// PLL is locked : it delivers a steady frequency.
+pub struct Locked;
+
+impl State for Disabled {}
+impl State for Locked {}
+impl State for Locking {}
+
+/// Trait to handle both underlying devices from the PAC (PLL_SYS & PLL_USB)
+pub trait PhaseLockedLoopDevice: Deref<Target = rp2040_pac::pll_sys::RegisterBlock> {}
+
+impl PhaseLockedLoopDevice for rp2040_pac::PLL_SYS {}
+impl PhaseLockedLoopDevice for rp2040_pac::PLL_USB {}
+
+/// A PLL.
+pub struct PhaseLockedLoop<S: State, D: PhaseLockedLoopDevice> {
+    device: D,
+    state: S
+}
+
+impl<S: State, D: PhaseLockedLoopDevice> PhaseLockedLoop<S, D> {
+    fn transition<To: State>(self, state: To) -> PhaseLockedLoop<To, D> {
+        PhaseLockedLoop {
+            device: self.device,
+            state: state
+        }
+    }
+
+    /// Releases the underlying device.
+    pub fn free(self) -> D{
+        self.device
+    }
+}
+
+
+/// Error type for the PLL module.
+/// See Chapter 2, Section 18 §2 for details on constraints triggering these errors.
+pub enum Error {
+
+    /// Proposed VCO frequency is out of range.
+    VCOFreqOutOfRange,
+
+    /// Feedback Divider value is out of range.
+    FBDIVOutOfRange,
+
+    /// Post Divider value is out of range.
+    PostDivOutOfRage,
+
+    /// Reference Frequency is out of range.
+    RefFreqOutOfRange,
+
+    /// Bad argument : overflows, bad conversion, ...
+    BadArgument
+}
+
+
+/// Parameters for a PLL.
+pub struct PLLConfig<R: Rate> {
+
+    /// Voltage Controlled Oscillator frequency.
+    pub vco_freq: R,
```

Comment:
I think I would put in the feedback divider value instead of the `vco_frequency`. Although then we need to carry around the XOsc frequency to calculate the effective output.